book-content/chapters/16-the-utils-folder.md

```diff
@@ -0,0 +1,1527 @@
+It's commonly thought that there are two levels of TypeScript complexity.
+
+On one end, you have library development. Here, you take advantage of many of TypeScript's most arcane and powerful features. You'll need conditional types, mapped types, generics, and much more to create a library that's flexible enough to be used in a variety of scenarios.
+
+On the other end, you have application development. Here, you're mostly concerned with making sure your code is type-safe. You want to make sure your types reflect what's happening in your application. Any complex types are housed in libraries you use. You'll need to know your way around TypeScript, but you won't need to use its advanced features much.
+
+This is the rule of thumb most of the TypeScript community use. "It's too complex for application code". "You'll only need it in libraries". But there's a third level that's often overlooked: the `/utils` folder.
+
+If your application gets big enough, you'll start capturing common patterns in a set of reusable functions. These functions, like `groupBy`, `debounce`, and `retry`, might be used hundreds of times across a large application. They're like mini-libraries within the scope of your application.
+
+Understanding how to build these types of functions can save your team a lot of time. Capturing common patterns means your code becomes easier to maintain, and faster to build.
+
+In this chapter we'll cover how to build these functions. We'll start with generic functions, then head to type predicates, assertion functions, and function overloads.
+
+## Generic Functions
+
+We've seen that in TypeScript, functions can receive not just values as arguments, but types too. Here, we're passing a _value_ and a _type_ to `new Set()`:
+
+```typescript
+const set = new Set<number>([1, 2, 3]);
+//                 ^^^^^^^^ ^^^^^^^^^
+//                 type     value
+```
+
+We pass the type in the angle brackets, and the value in the parentheses. This is because `new Set()` is a generic function. A function that can't receive types is a regular function, like `JSON.parse`:
```

Comment:
```suggestion
We pass the type in the angle brackets, and the value in the parentheses. This is because `new Set()` is a generic constructor function. A function that can't receive types is a regular function, like `JSON.parse`:
```